sql/log_event.h

```diff
@@ -2189,7 +2189,8 @@ class Query_log_event: public Log_event
     return !strncmp(query, "BEGIN", q_len) ||
       !strncmp(query, "COMMIT", q_len) ||
       !strncasecmp(query, "SAVEPOINT", 9) ||
-      !strncasecmp(query, "ROLLBACK", 8);
+      !strncasecmp(query, "ROLLBACK", 8) ||
+      !strncasecmp(query, C_STRING_WITH_LEN("XA "));
   }
```

Comment:
Ouch, yet another strcmp() added here :-(
Really, XA should be handled with a separate event like the XID_LOG_EVENT, not with yet another magic QUERY_LOG_EVENT that the code then needs to strcmp() against all over the place.

Reply:
> XA should be handled with a separate event like the XID_LOG_EVENT

Perhaps.
More useful I'd argue would be to first [check](https://github.com/MariaDB/server/blob/10.5-MDEV-33921/sql/log_event_server.cc#L1718)
**unlikely**  presence of replicate-filters and not to invoke
` is_trans_keyword()` altogether.

Reply:
In addition to @andrelkin 's suggestion, we can further optimize this with a mode to either strncmp for XA commands _or_ regular transaction commands (not both). The mode would be determined by `rgi->gtid_flags2 & (FL_PREPARED_XA | FL_COMPLETED_XA)`, so when true, we'd only do the `XA` check, and when false, we only check against regular transaction commands. I quickly prototyped a patch and it passed tests (though there is Galera and mysqlbinlog that also use this function which don't have immediate access to XA state, and may need additional extensions for them to make use of the optimization).